### PR TITLE
TestKeepsGoingUp - testDoesntWinWithWalls succeeds

### PR DESCRIPTION
**Pull request by gsingh62**

### CODE REVIEW

File: src/map/WorldMap.kt [hunk 1]
```diff
@@ -28,11 +30,13 @@ class CoordinateNodeWorldMap(private val nodes: MutableMap<Coordinate, Node>) :
         val newCoordinate = Coordinate(coordinate.x + deltax, coordinate.y + deltay)
         val oldPlace = nodes[coordinate]
         val newPlace = nodes[newCoordinate]
-        if (oldPlace is OpenSpaceNode && newPlace is OpenSpaceNode) {
+        if (newPlace is WallNode) {
+            throw UnsupportedOperationException("Hit wall node.")
```
Comment (alepar) [inline src/map/WorldMap.kt:34]: This feels like your own custom Exception type, eg InvalidMoveException

File: src/map/Node.kt [hunk 2]
```diff
@@ -9,11 +12,16 @@ class Actor(private val lifeUnits: Int) : MapObject {
 
 }
 
+class AlwaysGoUpPlayer(override val actor: Actor): Player {
```
Comment (alepar) [inline src/map/Node.kt:15]: why is this specific test implementation in production sources?

File: src/game/Game.kt [hunk 3]
```diff
@@ -2,31 +2,39 @@ package game
...
             timeCounter++
         }
     }
 
+    private fun moveNorthLogic() {
+        try {
+            worldMap.moveObject(player.actor, 0, -1)
+        } catch (ex: UnsupportedOperationException) {
+            when (ex.message) {
+                "Hit wall node." -> if (player is AlwaysGoUpPlayer) { player.receiveFeedback(WALL_FEEDBACK) }
```
Comment (alepar) [inline src/game/Game.kt:36]: why only that type of player gets notified of hitting the wall? i thought prohibiting going through the walls is a generic rule?

File: src/game/Game.kt [hunk 4]
```diff
@@ -2,31 +2,39 @@ package game
...
-                else -> {
-                    throw UnsupportedOperationException()
-                }
+            when (player.chooseNextMove()) {
+                Action.MOVE_NORTH -> moveNorthLogic()
             }
-
-            if (!player.actor.alive || worldMap.getNode(worldMap.positionFor(player.actor)) is ExitNode) {
+            if (!player.actor.alive ||
+                    player.feedback == WALL_FEEDBACK ||
```
Comment (alepar) [inline src/game/Game.kt:23]: I dont think the game should stop if the player tries to go through the wall. Just players position doesnt change, he gets the feedback, but the world goes on. Player should have a chance to correct his mistake (ie learn the concept of the wall on his own).
If you do implement it this way, however, you will have a problem of infinite game, but we could address this by adding a more generic terminate condition (global turn limit or player not moving for x turns etc)

File: src/game/Game.kt [hunk 5]
```diff
@@ -2,31 +2,39 @@ package game
 
 import action.Action
 import agent.Player
+import map.AlwaysGoUpPlayer
 import map.ExitNode
 import map.WorldMap
 import java.lang.UnsupportedOperationException
 
 class Game(private val player: Player, private val worldMap: WorldMap) {
+    companion object {
```
Comment (alepar) [inline src/game/Game.kt:11]: this feels more like enum. or exception type.

**Review by alepar — CHANGES_REQUESTED** [review]

File: src/game/Game.kt (hunk 5 above)
Comment (gsingh62) [reply src/game/Game.kt:11]: Yea moving this to its own class called Exception Messages

File: src/map/WorldMap.kt (hunk 1 above)
Comment (gsingh62) [reply src/map/WorldMap.kt:34]: agree

File: src/map/Node.kt (hunk 2 above)
Comment (gsingh62) [reply src/map/Node.kt:15]: my mistake, I wanted to use it for conditional exception handling - but we already discussed that didn't make sense

File: src/game/Game.kt (hunk 3 above)
Comment (gsingh62) [reply src/game/Game.kt:36]: discussed -this does not make sense - all players need feedback that they have hit a wall

File: src/game/Game.kt (hunk 4 above)
Comment (gsingh62) [reply src/game/Game.kt:23]: you're right will terminate after player is not alive anymore - to achieve this he has to lose 1 life unit per turn

**Comment by alepar** [comment]
close this PR please

**Comment by gsingh62** [comment]
done

**Pull request closed without merging**